osu.Game.Rulesets.Taiko/Objects/ISkinnableSwell.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Taiko.Objects.Drawables;
+using osu.Game.Skinning;
+
+namespace osu.Game.Rulesets.Taiko.Objects
+{
+    public interface ISkinnableSwell
+    {
+        void OnUserInput(DrawableTaikoHitObject<Swell> swell, int numHits, SkinnableDrawable mainPiece);
+
+        void OnHitObjectEnd(ArmedState state, SkinnableDrawable mainPiece);
+
+        /// <summary>
+        /// Applies passive transforms on HitObject start. Gets called every time DrawableTaikoHitobject
+        /// changes state. This happens on creation, and when the object is completed (as in hit or missed).
+        /// </summary>
+        void ApplyPassiveTransforms(DrawableTaikoHitObject<Swell> swell, SkinnableDrawable mainPiece);
```

Comment:
This interface design and naming feels very ad hoc (and I don't know what "passive transforms" are). If I were to suggest anything:

- `ApplyPassiveTransforms()` -> `AnimateSwellStart()`
- `OnUserInput()` -> `AnimateSwellProgress()`
- `OnHitObjectEnd()` -> `AnimateSwellCompletion()`

Better yet would be to merge `AnimateSwell{Start,Completion}()` into a single `AnimateState(ArmedState, SkinnableDrawable)` method.

Reply:
I think it's better to keep this in three methods, as combining them would require me to write or undo usable logic that already exists verbatim in virtual methods from the `DrawableHitObject` class:

`AnimateSwellStart()` is called from `UpdateStartTimeStateTransforms()`, which is overridden in the `DrawableSwell` class and applies a `BeginAbsoluteSequence` that makes transforms line up with the **hitobject start** ("passive transform").

`AnimateSwellCompletion()` is called from `UpdateHitStateTransforms()`, which is also overridden in `DrawableSwell` and applies a `BeginAbsoluteSequence` that makes transforms line up with state hit/miss aka the **hitobject end**.

Agree with the naming suggestions, have applied those.